crates/autopilot/src/database/on_settlement_event_updater.rs

```diff
@@ -0,0 +1,87 @@
+use {
+    anyhow::Context,
+    database::{
+        auction_prices::AuctionPrice,
+        byte_array::ByteArray,
+        settlement_observations::Observation,
+    },
+    ethcontract::{H160, U256},
+    number_conversions::u256_to_big_decimal,
+    std::collections::BTreeMap,
+};
+
+#[derive(Debug, Clone)]
+pub struct SettlementUpdate {
+    pub block_number: i64,
+    pub log_index: i64,
+    pub auction_id: i64,
+    pub tx_from: H160,
+    pub tx_nonce: i64,
+    pub gas_used: U256,
+    pub effective_gas_price: U256,
+    pub surplus: U256,
+    pub fee: U256,
+    pub prices: BTreeMap<H160, U256>,
+}
+
+impl super::Postgres {
+    pub async fn update_settlement_details(
+        &self,
+        settlement_update: SettlementUpdate,
+    ) -> anyhow::Result<()> {
+        let _timer = super::Metrics::get()
+            .database_queries
+            .with_label_values(&["update_settlement_details"])
+            .start_timer();
+
+        let mut ex = self.0.begin().await?;
+
+        // update settlements
+        database::auction_transaction::insert_settlement_tx_info(
+            &mut ex,
+            settlement_update.block_number,
+            settlement_update.log_index,
+            &ByteArray(settlement_update.tx_from.0),
+            settlement_update.tx_nonce,
+        )
+        .await
+        .context("insert_settlement_tx_info")?;
+
+        // update settlement_observations
+        database::settlement_observations::insert(
+            &mut ex,
+            Observation {
+                block_number: settlement_update.block_number,
+                log_index: settlement_update.log_index,
+                gas_used: u256_to_big_decimal(&settlement_update.gas_used),
+                effective_gas_price: u256_to_big_decimal(&settlement_update.effective_gas_price),
+                surplus: u256_to_big_decimal(&settlement_update.surplus),
+                fee: u256_to_big_decimal(&settlement_update.fee),
+            },
+        )
+        .await
+        .context("insert_settlement_observations")?;
+
+        // update auction_prices
+        database::auction_prices::delete(&mut ex, settlement_update.auction_id)
+            .await
+            .context("delete_auction_prices")?;
+        database::auction_prices::insert(
```

Comment:
Why do the prices need to be deleted first?

This function is guaranteed to only run once per settlement because `get_settlement_event_without_tx_info` will only run once if the tx info is successfully inserted and all the insertions happen in a transaction.

Edit: I see we also insert prices in `solvable_orders.rs`. I don't think we should. See my comment there.